### PR TITLE
Introducing concurrency configuration

### PR DESCRIPTION
**Pull request by messense**
When using async, you don't need lots of threads to achieve more concurrency.

10 threads 10 concurrency:

```
Threads 10
Iterations 1000
Rampup 0
Base URL http://localhost:9000


Fetch users               Total requests            10000
Fetch users               Successful requests       10000
Fetch users               Failed requests           0
Fetch users               Median time per request   2ms
Fetch users               Average time per request  2ms
Fetch users               Sample standard deviation 1ms

Concurrency Level         10
Time taken for tests      2.0 seconds
Total requests            10000
Successful requests       10000
Failed requests           0
Requests per second       5010.52 [#/sec]
Median time per request   2ms
Average time per request  2ms
Sample standard deviation 1ms
```

10 threads 50 concurrency:

```
Threads 10
Iterations 1000
Rampup 0
Base URL http://localhost:9000


Fetch users               Total requests            50000
Fetch users               Successful requests       50000
Fetch users               Failed requests           0
Fetch users               Median time per request   5ms
Fetch users               Average time per request  5ms
Fetch users               Sample standard deviation 2ms

Concurrency Level         50
Time taken for tests      5.1 seconds
Total requests            50000
Successful requests       50000
Failed requests           0
Requests per second       9792.15 [#/sec]
Median time per request   5ms
Average time per request  5ms
Sample standard deviation 2ms
```

### CODE REVIEW

File: SYNTAX.md [hunk 1]
```diff
@@ -26,6 +27,7 @@ plan:
 ### Benchmark main properties
 
 - `threads`: Number of threads running your plan concurrently. (Optional, default: 1)
+- `concurrency`: Number of concurrent requests. (Optional, default: 1)
```
Comment (messense) [inline SYNTAX.md:30]: I am not sure `concurrency` is the right term. `ab` uses `concurrency` where as `wrk` uses `connections`.

File: src/config.rs [hunk 2]
```diff
@@ -25,12 +26,14 @@ impl Config {
     let config_doc = &config_docs[0];
 
     let threads = read_i64_configuration(config_doc, "threads", NTHREADS);
+    let concurrency = read_i64_configuration(config_doc, "concurrency", threads);
```
Comment (messense) [inline src/config.rs:29]: Defaults to `threads` to avoid breaking change.

**Comment by fcsonline** [comment]
Thanks for this!

Yes, since we have async, `threads` is a misleading concept. I like the new `concurrency` parameter. I would warn somehow that `threads` is deprecated if it is used. This will let us remove it in the future. Having two similar parameters for the same things is not good.

The documentation should reflect that the `threads` parameter is deprecated too. On the other hand,  all samples in `examples` folder should be updated to use the new one.

File: SYNTAX.md (hunk 1 above)
Comment (fcsonline) [reply SYNTAX.md:30]: I like it. :ok_hand:

**Comment by messense** [comment]
> Having two similar parameters for the same things is not good.

Actually they are not really that similar, `threads` is for setting up `tokio`'s thread pool size, whereas `concurrency` is for setting up how many concurrent requests to make at a time.

I'd argue we should keep them both, but do not make `concurrency` default to `threads`. It's kind of like what `wrk` do with its `--threads` and `--connections`.

**Comment by fcsonline** [comment]
> Actually they are not really that similar, threads is for setting up tokio's thread pool size, whereas concurrency is for setting up how many concurrent requests to make at a time.

Ok, got it. What we can check too is that the number of threads is not higher than number of cores.

**Comment by fcsonline** [comment]
There is something not clear to me about this change.

Taking this benchmark file as an example, how many requests in total do you expect?
```yaml
---
threads: 1
concurrency: 5
base: 'http://localhost:9000'
iterations: 10

plan:
  - name: Fetch users
    request:
      url: /api/users.json
```

I get this output and stats:

```
Threads 1
Iterations 10
Rampup 0
Base URL http://localhost:9000


Fetch users               Total requests            50
Fetch users               Successful requests       50
Fetch users               Failed requests           0
Fetch users               Median time per request   2007ms
Fetch users               Average time per request  2009ms
Fetch users               Sample standard deviation 10ms

Concurrency Level         5
Time taken for tests      20.1 seconds
Total requests            50
Successful requests       50
```

Maybe you understand `concurrency` with another meaning, but I don't expect this parameter to change the total number of requests performed by drill if I change it.

Also, another thing we need to take in account is that the `plan` node can contain more than one request and those requests need to be performed sequentially because there can be data dependencies between requests. Does `concurrency` means concurrent iterations in this case?

I expect that the total number of requests performed by drill is (#iterations x #requests-in-plan-node). I think we have been influenced by `wrk` and `ab` ways to work to achieve the maximum throughput but a test plan is a list of requests dependent on each other.

**Comment by messense** [comment]
Before this patch, total requests equals `threads * iterations * count(plan nodes)`, with this patch it equals `concurrency * iterations * count(plan nodes)`, the only variable change is `threads` to `concurrency`.

**Comment by fcsonline** [comment]
> Before this patch, total requests equals threads * iterations * count(plan nodes), with this patch it equals concurrency * iterations * count(plan nodes), the only variable change is threads to concurrency.

Exactly, for this reason, they are quite similar. The only difference is that `threads` configuration can not exceed the number of cores in your machine. At least, it doesn't make sense. Regarding `concurrency` (number of concurrent iterations), this number can be above the number of cores.

For this reason, I was thinking about removing `threads` configuration in favor of `concurrency` and send to tokio's thread pool `min(#number_of_cores, #concurrency)` value. With this, we avoid the misconfiguration I mentioned before and on the other hand, we achieve maximum / the desired throughput.

**Comment by fcsonline** [comment]
Ok, now it looks much better! :clap:

Although, when you change the `concurrency` setting, it changes the number of total requests it does.  I think we can change this later because it needs a big refactor of how iterations are executed and how reports are collected. I'll implement a proposal.

Maybe we can merge this and move forward.

**Pull request merged**